util/src/main/java/com/google/cloud/hadoop/util/GoogleCredentialWithIamAccessToken.java

```diff
@@ -0,0 +1,73 @@
+package com.google.cloud.hadoop.util;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.client.auth.oauth2.TokenResponse;
+import com.google.api.client.googleapis.auth.oauth2.GoogleCredential;
+import com.google.api.client.http.HttpRequestInitializer;
+import com.google.api.client.http.HttpTransport;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.api.client.util.Clock;
+import com.google.api.services.iamcredentials.v1.IAMCredentials;
+import com.google.api.services.iamcredentials.v1.model.GenerateAccessTokenRequest;
+import com.google.api.services.iamcredentials.v1.model.GenerateAccessTokenResponse;
+import java.io.IOException;
+import java.time.Instant;
+
+public class GoogleCredentialWithIamAccessToken extends GoogleCredential {
+  private final String serviceAccountName;
+  private final HttpRequestInitializer initializer;
+  private final HttpTransport transport;
+  private final Clock clock;
+
+  private static final String DEFAULT_ACCESS_TOKEN_LIFETIME = "3600s";
+  private static final String DEFAULT_SERVICE_ACCOUNT_NAME_PREFIX = "projects/-/serviceAccounts/";
+
+  public GoogleCredentialWithIamAccessToken(
+      String serviceAccountName,
+      HttpRequestInitializer initializer,
+      HttpTransport transport,
+      Clock clock)
+      throws IOException {
+    this.serviceAccountName = serviceAccountName;
+    this.initializer = initializer;
+    this.transport = transport;
+    this.clock = clock;
+    initialize();
+  }
+
+  private void initialize() throws IOException {
+    GenerateAccessTokenResponse accessTokenResponse = generateAccessToken();
+    Instant expirationTimeInInstant = Instant.parse(accessTokenResponse.getExpireTime());
+    setAccessToken(accessTokenResponse.getAccessToken());
+    setExpirationTimeMilliseconds(expirationTimeInInstant.toEpochMilli());
+  }
+
+  @Override
+  protected TokenResponse executeRefreshToken() throws IOException {
+    GenerateAccessTokenResponse accessTokenResponse = generateAccessToken();
+    Instant expirationTimeInInstant = Instant.parse(accessTokenResponse.getExpireTime());
+    Long expirationTimeMilliSeconds = expirationTimeInInstant.toEpochMilli();
+    return new TokenResponse()
+        .setAccessToken(accessTokenResponse.getAccessToken())
+        .setExpiresInSeconds(
+            expirationTimeMilliSeconds == null
+                ? null
+                : (expirationTimeMilliSeconds - clock.currentTimeMillis()) / 1000);
+  }
+
+  private GenerateAccessTokenResponse generateAccessToken() throws IOException {
+    IAMCredentials IAMCredentials =
+        new IAMCredentials(transport, new JacksonFactory(), initializer);
+    GenerateAccessTokenRequest content = new GenerateAccessTokenRequest();
+    content.setScope(CredentialFactory.GCS_SCOPES);
```

Comment:
Is it necessary to hard-code scope? Can we use a value passes in `createScoped(...)` method instead?

Reply:
This parameter is for request made to IAM and is required referring to [documentation](https://cloud.google.com/iam/docs/creating-short-lived-service-account-credentials). createScoped() doesn't seem to do anything at the moment when creating Credentials, based on another [TODO:](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/f122baabed8c9a90a8fada5d1486b39ebdcfcab9/util-hadoop/src/main/java/com/google/cloud/hadoop/util/CredentialFromAccessTokenProviderClassFactory.java#L127) in the code.

Reply:
This TODO is outdated, `createScoped` was implemented 2 years ago:
https://github.com/googleapis/google-api-java-client/commit/b43c9caca959388849cadcb3b2dca26bcd246b2f#diff-3501428322606b937de4d7bdc68f66cb

Reply:
ok, obsolete comment deleted. Scope set on request to IAM is still separate and required regardless, scope in createScoped() is used with IAM access token to access GCS.

Reply:
Can we pass scope as a constructor parameter? This way this class can be used in BQ connector too - this class is in `utils` project already, so ideally we should make it more or less universal.

Reply:
Done.